src/inmanta/server/services/userservice.py

```diff
@@ -0,0 +1,125 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import logging
+
+import asyncpg
+
+import nacl.exceptions
+import nacl.pwhash
+from inmanta import const, data, protocol
+from inmanta.const import MIN_PASSWORD_LENGTH
+from inmanta.data import AuthMethod, model
+from inmanta.protocol import common, exceptions
+from inmanta.server import SLICE_DATABASE, SLICE_TRANSPORT, SLICE_USER
+from inmanta.server import config as server_config
+from inmanta.server import protocol as server_protocol
+
+LOGGER = logging.getLogger(__name__)
+
+
+def verify_authentication_enabled() -> None:
+    """raises an UnauthorizedException exception if server authentication is not enabled"""
+    if not server_config.server_enable_auth.get():
+        raise exceptions.BadRequest(
+            "Server authentication should be enabled. To setup the initial user use the inmanta-initial-user-setup tool."
+        )
+
+
+class UserService(server_protocol.ServerSlice):
+    """Slice for managing users"""
+
+    def __init__(self) -> None:
+        super().__init__(SLICE_USER)
+
+    def get_dependencies(self) -> list[str]:
+        return [SLICE_DATABASE]
+
+    def get_depended_by(self) -> list[str]:
+        return [SLICE_TRANSPORT]
+
+    @protocol.handle(protocol.methods_v2.list_users)
+    async def list_users(self) -> list[model.User]:
+        return sorted([user.to_dao() for user in await data.User.get_list()], key=lambda u: u.username)
+
+    @protocol.handle(protocol.methods_v2.add_user)
+    async def add_user(self, username: str, password: str) -> model.User:
+        verify_authentication_enabled()
+        if not password or len(password) < MIN_PASSWORD_LENGTH:
+            raise exceptions.BadRequest("the password should be at least 8 characters long")
+
+        # hash the password
+        pw_hash = nacl.pwhash.str(password.encode())
+
+        # insert the user
+        try:
+            user = data.User(
+                username=username,
+                password_hash=pw_hash.decode(),
+                auth_method=AuthMethod.database,
+            )
+            await user.insert()
+        except asyncpg.UniqueViolationError:
+            raise exceptions.Conflict(f"A user with name {username} already exists.")
+        return user.to_dao()
+
+    @protocol.handle(protocol.methods_v2.delete_user)
+    async def delete_user(self, username: str) -> None:
+        verify_authentication_enabled()
+        user = await data.User.get_one(username=username)
+        if user is None:
+            raise exceptions.NotFound(f"User with name {username} does not exist.")
+
+        await user.delete()
+
+    @protocol.handle(protocol.methods_v2.set_password)
+    async def set_password(self, username: str, password: str) -> None:
+        verify_authentication_enabled()
+        if not password or len(password) < MIN_PASSWORD_LENGTH:
```

Comment:
We should all this constraint on the password length to the API documentation. The front-end, for example, will find that information rather useful.